plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
+        $visitFieldsToSelect = array_merge(array('idvisit', 'location_ip'), array_keys(self::$logVisitFieldsToUpdate));
+
+        $lastId = 0;
+        do {
+            $logs = $this->dao->getVisitsWithDatesLimit($from, $to, $visitFieldsToSelect, $lastId, $segmentLimit);
+            if (!empty($logs)) {
+                $lastId = $logs[count($logs) - 1]['idvisit'];
+
+                $this->reattributeVisitLogs($output, $logs);
+            }
+        } while (count($logs) == $segmentLimit);
+    }
+
+    protected function reattributeVisitLogs(OutputInterface $output, $logRows)
+    {
+        foreach ($logRows as $row) {
+            if (!$this->isRowComplete($output, $row)) {
+                continue;
+            }
+
+            $location = $this->getVisitLocation($row);
+            $valuesToUpdate = $this->getValuesToUpdate($row, $location);
+
+            $this->onVisitProcessed($output);
+
+            $idVisit = $row['idvisit'];
+
+            if (empty($valuesToUpdate)) {
+                $this->writeIfVerbose($output, 'Nothing to update for idvisit = ' . $idVisit . '. Existing location info is same as geolocated.');
+            } else {
+                $this->writeIfVerbose($output, 'Updating visit with idvisit = ' . $idVisit . '.');
+
+                $this->dao->updateVisits($valuesToUpdate, $idVisit);
+                $this->dao->updateConversions($valuesToUpdate, $idVisit);
+            }
+        }
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return int
+     */
+    protected function getPercentStep(InputInterface $input)
+    {
+        $percentStep = $input->getArgument(self::PERCENT_STEP_ARGUMENT);
+
+        if (!is_numeric($percentStep)) {
+            return self::PERCENT_STEP_ARGUMENT_DEFAULT;
+        }
+
+        // Percent step should be between maximum percent value and minimum percent value (1-100)
+        if ($percentStep > 99 || $percentStep < 1) {
+            return 100;
+        }
+
+        return $percentStep;
+    }
+
+    /**
+     * Returns location log values that are different than the values currently in a log row.
+     *
+     * @param array $row The visit row.
+     * @param array $location The location information.
+     * @return array The location properties to update.
+     */
+    protected function getValuesToUpdate(array $row, $location)
```

Comment:
> We do not only create a class to have it reusable. This was not my point here. As said I prefer to have a class for better readable code, for clean code and for testable code.

I understand the desire to refactor code into small, testable units, but all code is not created equal.

Classes that are never re-used (by developers or other code), that do not belong to an important sub-system, and are not used by more than one other class are just clutter (to me). (Other languages mitigate this issue w/ features like inner classes, package private classes or anonymous classes.)

In the context, of a small piece of code, like a plugin, I understand and agree that this type of organization is desired. In a large code base like Piwik, the extra class may not add to clarity, UNLESS it is properly abstracted. Which, in this pull request's case, would mean adding to core.

Now, as I said, I will move the visitor re-attribution logic, but it makes no sense to move the do-loop since it is only there for the command. To properly abstract this logic would mean an addition to core, and if this is actually desired, I can do it in another pull request.

> Matt explained it quite well. It's not super important to me and it was only a suggestion. It's nothing were I would insist on something and to me the benefit is not worth a huge discussion.

I find it very rude to argue w/ someone and never try to reach consensus or attempt an understanding of the other's point of view. It may not be a big deal to you whether I make this change or not, but it is a big deal to me that we, as team members, either come to a consensus or understand each other's differing position.

> I think a client should have something where we can be pretty sure that it actually works.

As I said there are accuracy tests, and I can improve them. Unit tests for refactored (which I could easily add AFTER merging this pull request and before the release) won't test more than the existing system test, because the logic for this command is SIMPLE.

Additionally, the code I'm working on is not as central or complicated as the Sort filter, so the comparison is invalid.

You might notice if you look at the command, ignore output statements and think about the context in which the non-method calls occur.

> Also we most likely won't add unit/integration tests once a feature is "done" and shipped ;)

I maintain my own TODO lists for work that should be done later after other prioritized work.